### PR TITLE
Week07

### PR DESCRIPTION
**Pull request by wolves**
Summary
---------------------------------------------------
*Summarize what is included in this PR?*

This pull request contains code for a production line monitored by a manager that starts up with the provided count of workers and returns completed products based on the number of products provided. The manager can be provided a context for cancellation if necessary. This functionality is included alongside a test suite providing 97.7% test coverage.

Code Description
------------------------------------------
*Briefly describe what your code intends to do*

This week's code is an iteration of last week's code with the addition of contexts and a `Run` function that lives in the manager file. With the addition of `context.Context` the ability to provide user-signaled interrupts as well as timeouts has been added.

The `Run` function takes in the provided context, starts up the worker pool passing along the context with it, assigns the provided products, and then watches for either the completed product list, errors via the manager's error channel, or the context.Done() output. 

The context that is sent to the `Manager.Start()` method is passed on to the `Employee.work()` method which also watches for the context cancellation.

Essay/Commentary
------------------------------------------
*A long-form writeup covering: Difficulties encountered / Noteworthy choices made / Other things that stood out to you*

This week's assignment gave me a run for my money. Achieving the required test coverage was relatively straight forward but I spent more than a comfortable amount of time wrestling with race conditions. I was able to achieve 100% coverage at one point but not without encountering race conditions that I was unable to fix. I chose to pull out some of the logic in my code that was checking contexts and passing around Goroutines to do so and solve my race conditions while maintaining a respectable test coverage percentage. 

From the Thursday class, I noted that I should be looking to remove the use of the `Manager.Done` method but after multiple different attempts, I was not able to achieve this without causing my tests to permanently block or also re-introduce further race conditions. I look forward to seeing this refactor and learning the correct implementation at the next class session.

As a whole, I was able to achieve the required assignment criteria. That being said, I can't help but feel that I am not completely content with the resulting product. I refactored several times only to feel as though I was overthinking things and less and less confident in the behavior of the code. This may be a sign that I need to continue exercising my understanding of contexts as well as channels.

Finally, upon pushing my code to Github, my test suite informed me that the build for windows was failing. I utilized the logs to determine that Windows was returning `undefined` for `syscall.SIGUSR2` as-well-as `syscall.Kill` which I was using in my tests for the user interrupt functionality. Having followed this week's reading example to formulate this test I was a bit surprised, but feeling a bit bewildered already with my overall submission, I was not about to let this be the cherry on top of my disappointment sundae.

I took a look at the `syscall` documentation and found `syscall.SIGINT` which felt a bit more like what I was testing anyway so I substituted that and after pushing to GitHub, was happy to see that my `undefined` woes were halved leaving me with just `syscall.Kill` to handle.

I searched for this issue online and discovered that there seemed to be a known issue involving Go's `syscall.Kill` when building for windows. I was able to substitute that test call with the following:

```Go
const TEST_SIGNAL = syscall.SIGINT

p, _ := os.FindProcess(os.Getpid())
p.Signal(TEST_SIGNAL)
```

This addressed the `undefined` errors I was seeing that were thwarting my tests but in turn caused my test suite not to recognize the user interrupt signal on the windows build. 

I am unable to figure out how to get the Windows build to properly recognize a user interrupt beyond what I have included. I hope that in the case this is an oversight in this week's reading, we can address it in Monday's class so I will have that knowledge moving forward.

Overall, there are a few things I'm taking away from this week's assignment. First, I plan to exert a good amount of time beyond the class finding much more comfort with the `context` package as it is an important one and I have room to grow. Second, concurrency, whenever implemented, will always require a great deal of close attention to properly execute, and finally, just because code criteria are satisfied in my local test suite does not necessarily guarantee that to be the case in all areas of implementation. Being aware of that and verifying it via a full cross-platform test environment will help ensure better-resulting code as a whole.

I look forward to going over some of my outstanding questions and uneasiness in the next class so that as I continue on my journey as a Go developer I can find more confidence concerning some of this week's hiccups.

### CODE REVIEW

File: week07/completed_product_test.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+		expErr bool
+	}{
+		{name: "a valid product is IsValid", stubs: validProduct, expErr: false},
+		{name: "no employee product IsInvalid", stubs: noEmpProd, expErr: true},
+		{name: "no quantity product IsInvalid", stubs: noQtyProd, expErr: true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			p := &tt.stubs.Product
+			_ = p.Build(tt.stubs.Employee)
```
Comment (markbates) [inline week07/completed_product_test.go:32]: Always check your errors.

File: week07/manager_test.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+		_, err := Run(context.Background(), e, ip...)
+		if err == nil {
+			t.Fatal("expected error received none")
+		}
+
+	})
+
+	// TODO: interruption by a signal
+	t.Run("with interrupt signal", func(t *testing.T) {
+		const TEST_SIGNAL = syscall.SIGINT
```
Comment (markbates) [inline week07/manager_test.go:153]: Try using `SIGUSR1` to fix the windows problem. If that doesn't work. mark the test with `t.Skip()` so it doesn't keep failing in the future. It's OK to skip this test and move on.

```suggestion
		const TEST_SIGNAL = syscall.SIGUSR1
```

**Review by markbates — COMMENTED** [review]
Excellent work!

File: week07/manager_test.go (hunk 2 above)
Comment (wolves) [reply week07/manager_test.go]: `SIGUSR1` came back undefined, so skipping and commenting out the test. Thank you for the feedback

**Pull request merged**